### PR TITLE
feat(chore): add lint using CI

### PR DESCRIPTION
**Pull request by mvorisek**
## Description

Unified CS improves readability, prevents hidden mistakes and enforce unified contribution standards. And in a long term, it also simplifies merge conflicts.

The CS rules are based on industry standard `eslint-config-airbnb-base`, but all rules might be too hard to fix at once. This PR focuses on whitespace CS which can be autofixed /wo having to review each line manually.

related https://github.com/fomantic/Fomantic-UI/issues/29

nicely integrated with Github Actions like:

image

### CODE REVIEW

**Comment by prudho** [comment]
Yes, code linting is definitely something we need to implement. I started using myself eslint at work and this is something i don't regret at all. But... 😅 

We cannot implement it, and already use it to enforce merge requests. I think that first we need to define which rules we want to use and the level of error it'll raise. Then we can do some merge requests to fix the code. But we can't fix all the code in one PR, maybe a PR per component.

Once this is done, then we can enforce lint check trough the CI pipeline.

**Comment by mvorisek** [comment]
I have updated the PR, instead of using `airbnb-base` as a base eslint config (with possible side effects) I added the most important whitespaces rules only to match general coding standards. With whitespace changes the code behaviour is not changes and this PR is safe to merge.

In another PR, more eslint rules can be added.

Also don't worry about merge conflicts (from changes based on code before this PR), they can be easily resolved, just cherry pick the eslint rules and run `npm run lint-fix`, you can then integrate the changes as usual.

**Review by lubber-de — COMMENTED** [review]
As the PR is basically ok now, i just comment instead of requesting changes as i would like to hear opinion of other maintainers. See my comments below

File: examples/assets/library/jquery.min.js
Comment (mvorisek) [reply examples/assets/library/jquery.min.js]: fixed - `/examples/assets/library` added to eslint excluded files, no foreign library should be transformed

File: .github/workflows/ci.yml
Comment (mvorisek) [reply .github/workflows/ci.yml:2]: Having here `push` implicitly is done in a lof of project, any push will run CI.

This makes experiments easier and also runs CI for development/master and other release branches.

If ok with you, I prefer to keep `push`.

File: .eslintrc.js
Comment (mvorisek) [reply .eslintrc.js:18]: I personally highly prefer 4 spaces as 2 spaces requires more eye stress to track the current indentation level and also forces you visually to not write deeply nested code /wo using methods.

So when normalizing the CS now, I propose 4 spaces. But if most other maintainers will require 2 spaces, I would not/cannot object.

**Comment by lubber-de** [comment]
I took a look at the complete files to get myself convinced about the 4 indent 😉  and  i recognized the linter does sometimes not implement the 4 indent correctly, which
needs (manual?) adjustments, otherwise the new indent is making it worse than before at those codelines, because it gets confusing

For example dimmer.js, look at the semicolon, which terminates the var statement
https://github.com/fomantic/Fomantic-UI/blob/9074205244996ae83a06273e5364ca2be190fca4/src/definitions/modules/dimmer.js#L38

It is now looking like that
image

while it was and is supposed to look like that (same indent as the var statement)
image

**Comment by mvorisek** [comment]
Maybe we need another eslint rule for semicolon, I will have a look on a solution for it.

**Comment by mvorisek** [comment]
The solution is to run eslint `semi: ['error', 'never']` once and then again. There is no explicit config of indentation for semicolon in eslint.

**Comment by lubber-de** [comment]
Well, when adding the semicolon to the end of the last line, why the indention at all then (in such cases)?

image


Also, you added the rule to add a trailing comma in object literals. The linter idea behind that is to easily exchange/adding/removing single lines without the need to also add another comma to the previous attribute notation.
This is the same for such use cases where the semicolon has its own line.

So: Adding the trailing comma but also adding a semicolon to the last entry is not consistent anymore (IMHO even worse)
I would really prefer to have the semicolon being indented as it was before. This clearly shows where the indentation block ends as it is now.

Now we have the discussion i basically want to avoid ;) And now imagine new contributors trying to help and we are going to argue about "correct" whitespacing :) 

Don't get me wrong, I am not against the linting. But indentin the semicolons would be the cleaner approach. Yes, this possibly needs a (onetime) manual adjustment if there is no 100% automatic solution.

**Comment by mvorisek** [comment]
> Well, when adding the semicolon to the end of the last line, why the indention at all then (in such cases)?

chained method call indentation is defacto standard (at least has always been in all codebase I have worked with) and it makes sense, as

```
$module
    .callOne()
    .callTwo();
$module.callOne();
```

is IMHO visually more readable and more compact than

```
$module
.callOne()
.callTwo()
;
$module.callOne();
```

> Don't get me wrong, I am not against the linting. But indentin the semicolons would be the cleaner approach. Yes, this possibly needs a (onetime) manual adjustment if there is no 100% automatic solution.

You are right here, I have reviewed the changes and there are at least two usecases with semicolon on a NL:

a) ternary:

image

I have never seen semicolon on NL for them, also no reason for easier diffing, as ternary operator is always binary.

b) var declaration:

image

I will fix these manually 👍

**Comment by exoego** [comment]
For the record, I am positive on introducing linter.
I believe it helps both contributors and maintainers.
I defer to lubber on styling preference since he is the most active maintainer.

**Comment by lubber-de** [comment]
> ```
> $module
>     .callOne()
>     .callTwo();
> $module.callOne();
> ```
> 
> is IMHO visually more readable and more compact than
> 
> ```
> $module
> .callOne()
> .callTwo()
> ;
> $module.callOne();
> ```
> 

My point basically was about comparison to the trailing comma rule.
If you want to add a .callThree() you now have to remove the semicolon from .callTwo() as well

> I will fix these manually 👍

Thx 😊

**Comment by mvorisek** [comment]
This PR is complete.

I have updated the PR title/description to stress it focuses solely on whitespaces. There is no single functional change.

I have reviewed all manual changes by generating unminified `semantic.js` before and after this PR and then applied eslint with the following <details>
  <summary>config</summary>
```
module.exports = {
    env: {
        browser: true,
        es6: true,
    },
    extends: [
    ],
    parserOptions: {
        ecmaVersion: '2020',
        sourceType: 'module',
    },
    ignorePatterns: [
        '/dist',
        '/examples/assets/library',
        '/test/coverage',
        '/test/helpers',
    ],
    rules: {
        indent: ['error', 4, { SwitchCase: 1 }],
        'keyword-spacing': ['error'],
        'space-in-parens': ['error'],
        'comma-spacing': ['error'],
        'space-before-function-paren': ['error', {
            'named': 'never',
            'anonymous': 'always',
            'asyncArrow': 'always',
        }],
        'space-infix-ops': ['error'],
        'padding-line-between-statements': ['error', {
            blankLine: 'always',
            prev: '*',
            next: ['continue', 'break', 'export', 'return', 'throw'],
        }],
        'spaced-comment': ['error', 'always', {
            line: {
                markers: ['/'],
                exceptions: ['-', '+'],
            },
            block: {
                markers: ['!'],
                exceptions: ['*'],
                balanced: true,
            },
        }],
        'padded-blocks': ['error', 'never'],
        'no-multiple-empty-lines': ['error', {
            max: 0,
            maxBOF: 0,
            maxEOF: 0,
        }],
        'no-multi-spaces': ['error'],
        'key-spacing': ['error'],
        'brace-style': ['error'],
        curly: ['error'],
        quotes: ['error', 'single', { avoidEscape: true }],
        'comma-dangle': ['error', {
            arrays: 'always-multiline',
            objects: 'always-multiline',
            functions: 'never',
            imports: 'always-multiline',
            exports: 'always-multiline',
        }],
        semi: ['error', 'never'],
        'no-extra-semi': ['error'],
        'no-trailing-spaces': ['error'],
    },
    reportUnusedDisableDirectives: true,
};
```
</details>

to remove all autofixed whitespace differencies as much as possible. The changes made look correct.

**Comment by lubber-de** [comment]
Thanks for doing this.
I am sorry to say, but i still don't agree to the inconsisty of having trailing commas in object literals, but also putting the semicolon to the end of  chained methods. What one solves is re-invented by the other

The idea for the trailing comma rule is
```javascript

// bad code as per lint rule
// When i want to add another object property, i have to change 2 lines (add the comma to c: 2 and my new line)
var a = {
    b: 1,
    c: 2
}

// 2 lines changed. Bad practice as per eslint rule
var a = {
    b: 1,
    c: 2,
    d: 3
}

//  but lint rule wants to only change one line which is considered goot practice
var a = {
    b: 1,
    c: 2,
    d: 3,
}
```

I am basically totally fine with the above rule :)

But now:
You changed from 
```javascript

$module
    .on('bla',function(){})
    .on('bla',function(){})
    .on('bla',function(){})
;
```

to

```javascript

$module
    .on('bla1',function(){})
    .on('bla2',function(){})
    .on('bla3',function(){});
```

Regardless of the opinion based readability, we now have invented the same bad practice problem as the trailing comma force was solving by the previous rule before.
If i want to add another chain function to the end of the chain i have to remove the semicolon in the previous line.

```javascript
// 2 lines need  to be changed
$module
    .on('bla1',function(){})
    .on('bla2',function(){})
    .on('bla3',function(){})
    .on('bla4',function(){});


// while the current code would allow for only adding the additional chain line
// 2 lines need  to be changed
$module
    .on('bla1',function(){})
    .on('bla2',function(){})
    .on('bla3',function(){})
    .on('bla4',function(){})
;

```

That's what i meant by inconsisty.

File: src/definitions/behaviors/form.js
Comment (lubber-de) [inline src/definitions/behaviors/form.js:799]: indent is completely missing

Comment (mvorisek) [reply src/definitions/behaviors/form.js]: fixed - eslint does not indent lines starting/ending with an operator

**Comment by mvorisek** [comment]
I was able to craft regexes to fix semicolon on a NL for chained calls. Is there anything left?

**Comment by mvorisek** [comment]
PR is done, I added more whitespaces rules to match AirBnB CS (basically the rules are copied from their ruleset). The hash of minified `semantic.min.js` file before and after this PR is unchanged, ie. there is no functional change.

Later, I plan to create another PR which will fix `var` -> `let` etc., which can have side effects.

**Review by lubber-de — APPROVED** [review]
LGTM! Thanks for doing this and your patience 😉

**Pull request merged**
